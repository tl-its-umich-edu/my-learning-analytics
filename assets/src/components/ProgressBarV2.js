import React from 'react'
import Line from './Line'
import Label from './Label'
import { withStyles } from '@material-ui/core/styles'
import { roundToXDecimals } from '../util/math'

const styles = theme => ({
  gradedBar: {
    backgroundColor: theme.palette.secondary.main
  },

  ungradedBar: {
    backgroundColor: theme.palette.info.main
  },

  outOfBar: {
    backgroundColor: theme.palette.negative.main
  }
})

function ProgressBarV2 (props) {
  const {
    classes,
    score,
    outOf,
    submitted,
    percentWidth,
    height = 10,
    margin,
    lines = [],
    description,
    onBarFocus,
    onBarBlur
  } = props

  const scoreRatio = score
    ? score / outOf
    : null

  const calculateLineLeftOffset = (value, maxValue) => {
    if (value > maxValue) return 100
    if (value < 0) return 0
    return value / maxValue * 100
  }

  const calculateLableValue = (value, maxValue) => {
    if (value > maxValue) return 100
    if (value < 0) return 0
    return value
  }

  return (
    <>
      {
        <div
          tabIndex='0'
          aria-label={description !== undefined ? description : undefined}
          className={submitted && !score ? classes.ungradedBar : classes.outOfBar} // submitted assignment, and ungraded yet : otherwise
          style={{
            width: `${percentWidth}%`,
            height: `${height}px`,
            position: 'relative',
            margin: margin
          }}
          onFocus={(event) => { if (onBarFocus) onBarFocus(event.currentTarget) }}
          onBlur={(event) => { if (onBarBlur) onBarBlur() }}
        >
          {
            lines.length > 0
              ? lines.filter(line => line.value || line.value === 0) // filter any lines without value
                .map((line, key) => (
                  <div key={key}>
                    <Line
                      height={height}
                      left={`${calculateLineLeftOffset(line.value, outOf)}%`}
                      color={line.color}
                      labelPlacement={line.labelPlacement}
                    />
                    {
                      line.label
                        ? (
                          <Label
                            value={`${calculateLableValue(line.value, outOf)}`}
                            color={line.color}
                            labelText={`${line.label}: ${roundToXDecimals(line.value, 1)}%`}
                            labelPlacement={line.labelPlacement}
                          />
                        )
                        : null
                    }
                  </div>
                ))
              : null
          }
          {
            scoreRatio
              ? (
                <div
                  className={classes.gradedBar}
                  style={{
                    position: 'absolute',
                    display: 'inline-block',
                    width: `${(scoreRatio * 100) > 100 ? 100 : scoreRatio * 100}%`,
                    height: `${height}px`
                  }}
                />
              )
              : null
          }
        </div>
      }
    </>
  )
}

export default withStyles(styles)(ProgressBarV2)
