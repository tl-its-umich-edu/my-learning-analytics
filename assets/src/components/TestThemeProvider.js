import * as React from 'react'
import { ThemeProvider } from '@material-ui/core/styles'

import { siteTheme } from '../globals'

function TestThemeProvider ({ children }) {
  return <ThemeProvider theme={siteTheme}>{children}</ThemeProvider>
}

export default TestThemeProvider
