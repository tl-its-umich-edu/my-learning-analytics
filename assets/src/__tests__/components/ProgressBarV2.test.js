/* global it, expect */

import React from 'react'
import ProgressBarV2 from '../../components/ProgressBarV2'
import renderer from 'react-test-renderer'
import TestThemeProvider from '../../components/TestThemeProvider'
it('renders correctly', () => {
  const tree = renderer
    .create(
      <TestThemeProvider>
        <ProgressBarV2
          submitted=''
          score={85}
          outOf={100}
          percentWidth={90}
          lines={[{
            label: 'Current',
            value: 80,
            color: 'steelblue',
            labelPlacement: 'down'
          },
          {
            label: 'Goal',
            value: 40,
            color: 'green',
            labelPlacement: 'up'
          },
          {
            label: 'Max Possible',
            value: 23,
            color: 'grey',
            labelPlacement: 'downLower'
          }]}
        />
      </TestThemeProvider>
    ).toJSON()

  expect(tree).toMatchSnapshot()
})
