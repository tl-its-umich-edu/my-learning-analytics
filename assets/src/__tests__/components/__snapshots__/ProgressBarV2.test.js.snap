// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ProgressBarV2-outOfBar-2"
  style={
    Object {
      "height": "10px",
      "margin": "0px",
      "position": "relative",
      "width": "90%",
    }
  }
>
  <div>
    <div
      style={
        Object {
          "backgroundColor": "steelblue",
          "display": "inline-block",
          "height": "35px",
          "left": "80%",
          "marginBottom": "-25px",
          "marginTop": 0,
          "position": "absolute",
          "width": "2px",
          "zIndex": 1,
        }
      }
    />
    <div
      style={
        Object {
          "color": "steelblue",
          "display": "inline-block",
          "left": "calc(80.5% - 115px)",
          "position": "absolute",
          "top": "55px",
          "zIndex": 1,
        }
      }
    >
      Current: 80%
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "backgroundColor": "green",
          "display": "inline-block",
          "height": "35px",
          "left": "40%",
          "marginBottom": 0,
          "marginTop": "-25px",
          "position": "absolute",
          "width": "2px",
          "zIndex": 1,
        }
      }
    />
    <div
      style={
        Object {
          "color": "green",
          "display": "inline-block",
          "left": "40.5%",
          "position": "absolute",
          "top": "-25px",
          "zIndex": 1,
        }
      }
    >
      Goal: 40%
    </div>
  </div>
  <div>
    <div
      style={
        Object {
          "backgroundColor": "grey",
          "display": "inline-block",
          "height": "52px",
          "left": "23%",
          "marginBottom": "-25px",
          "marginTop": 0,
          "position": "absolute",
          "width": "2px",
          "zIndex": 1,
        }
      }
    />
    <div
      style={
        Object {
          "color": "grey",
          "display": "inline-block",
          "left": "23.5%",
          "position": "absolute",
          "top": "75px",
          "zIndex": 1,
        }
      }
    >
      Max Possible: 23%
    </div>
  </div>
  <div
    className="ProgressBarV2-gradedBar-1"
    style={
      Object {
        "display": "inline-block",
        "height": "10px",
        "position": "absolute",
        "width": "85%",
      }
    }
  />
</div>
`;
